### ISSUE TITLE
[BUG] Number of trainable weights seem to change after model compilation

### ISSUE THREAD

**Issue by alpapado** [fchollet/keras#8121]
Consider the following scenario

```python
from keras.layers import Input, Dense
from keras.models import Model
from keras.optimizers import Adam

x = Input(shape=(100,))
y1 = Dense(units=32)(x)
model1 = Model(inputs=x, outputs=y1)
print("MODEL 1")
model1.summary()

model1.trainable = False
x = Input(shape=(100,))
y1 = model1(x)
y2 = Dense(units=64)(y1)
model2 = Model(inputs=x, outputs=y2)
model2.compile(optimizer=Adam(), loss='categorical_crossentropy')

print("MODEL 2")
model2.summary()

model1.trainable = True
print("MODEL 2 after")
model2.summary()
```

I would expect that, since `model2` has been compiled, the output of the two `model2.summary()` calls would be the same. However, after running the above code the actual output is

```
_________________________________________________________________
MODEL 2
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         (None, 100)               0         
_________________________________________________________________
model_1 (Model)              (None, 32)                3232      
_________________________________________________________________
dense_2 (Dense)              (None, 64)                2112      
=================================================================
Total params: 5,344
Trainable params: 2,112
Non-trainable params: 3,232
_________________________________________________________________
MODEL 2 after
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         (None, 100)               0         
_________________________________________________________________
model_1 (Model)              (None, 32)                3232      
_________________________________________________________________
dense_2 (Dense)              (None, 64)                2112      
=================================================================
Total params: 5,344
Trainable params: 5,344
Non-trainable params: 0
```

Is this the expected behavior? 

In addition, from other experiments I get the feeling that although the number of trainable weights reported by `summary()` differs in the two cases, the actual number of trainable weights is the expected one, that is, in the example above, even after setting `model1.trainable=True`, training will really update only 2,112 parameters and not 5,344. So, maybe this is just a reporting issue.

**Comment by RohithKuditipudi** [comment]
I get a very similar issue when running the following snippet from the Keras docs. Based on the output of summary(), it seems like even after compiling, the number of trainable parameters changes in both trainable_model and frozen_model depending on whether I set layer.trainable = True or layer.trainable = False.

x = Input(shape=(32,))
layer = Dense(32)
layer.trainable = False
y = layer(x)

frozen_model = Model(x, y)
# in the model below, the weights of `layer` will not be updated during training
frozen_model.compile(optimizer='rmsprop', loss='mse')

layer.trainable = True
trainable_model = Model(x, y)
# with this model the weights of the layer will be updated during training
# (which will also affect the above model since it uses the same layer instance)
trainable_model.compile(optimizer='rmsprop', loss='mse')
